shadowsocks/tcp.go

```diff
@@ -106,24 +106,36 @@ func findAccessKey(clientConn onet.DuplexConn, cipherList CipherList) (string, o
 		}
 		// Move the active cipher to the front, so that the search is quicker next time.
 		cipherList.SafeMarkUsedByClientIP(entry, clientIP)
-		ssr := NewShadowsocksReader(io.MultiReader(bytes.NewReader(replayBytes), clientConn), cipher)
+		ssr := NewShadowsocksReader(io.MultiReader(bytes.NewReader(firstBytes), clientConn), cipher)
 		ssw := NewShadowsocksWriter(clientConn, cipher)
-		return id, onet.WrapConn(clientConn, ssr, ssw).(onet.DuplexConn), nil
+		return id, onet.WrapConn(clientConn, ssr, ssw).(onet.DuplexConn), salt, nil
 	}
-	return "", clientConn, fmt.Errorf("Could not find valid TCP cipher")
+	return "", clientConn, nil, fmt.Errorf("Could not find valid TCP cipher")
 }
 
 type tcpService struct {
-	listener    *net.TCPListener
+	listener *net.TCPListener
+	// `ciphers` is a pointer to SSPort.cipherList, which can be updated by SSServer.loadConfig.
 	ciphers     *CipherList
 	m           metrics.ShadowsocksMetrics
 	isRunning   bool
 	readTimeout time.Duration
+	replayCache ReplayCache
 }
 
+// Prevent replays of this many of the most recent handshakes.
+// The false positive probability will be 2 * replayHistory / 2^32 = 1/200,000.
+const replayHistory = 10_000
+
 // NewTCPService creates a TCPService
 func NewTCPService(listener *net.TCPListener, ciphers *CipherList, m metrics.ShadowsocksMetrics, timeout time.Duration) TCPService {
-	return &tcpService{listener: listener, ciphers: ciphers, m: m, readTimeout: timeout}
+	return &tcpService{
+		listener:    listener,
+		ciphers:     ciphers,
+		m:           m,
+		readTimeout: timeout,
+		replayCache: NewReplayCache(replayHistory),
```

Comment:
This is creating a replayCache for each port.
I think this should be at SServer instead, and shared across the port.tcpServices, like the metrics, so servers with multiple ports don't get bloated. We have some long running servers, some may have dozens of ports.

Another thing in my mind: It should be perfectly fine for two different keys to use the same salt, since the encrypted content will be different anyway. Should we be isolating the different keys?

I'm also concerned about changing the behavior of existing servers. There should be at least a escape hatch. Let's have a flag with the value for the `replayHistory`, so server owners can disable if they need to. Or should it be a config option?

This can be useful for a gradual rollout. Let's discuss how we'll roll this out in a safe way.

Reply:
> This is creating a replayCache for each port.
> I think this should be at SServer instead, and shared across the port.tcpServices, like the metrics

Done.

> Another thing in my mind: It should be perfectly fine for two different keys to use the same salt, since the encrypted content will be different anyway. Should we be isolating the different keys?

This seems like a pathological client behavior, but I've addressed it by mixing the key ID into the map key, so that the same salt on different keys is unlikely to collide in the map.

> I'm also concerned about changing the behavior of existing servers. There should be at least a escape hatch. Let's have a flag with the value for the `replayHistory`, so server owners can disable if they need to. Or should it be a config option?

OK, I added a flag.